src/wayland-eglsurface.c

```diff
@@ -2564,26 +2517,45 @@ static EGLBoolean wlEglDestroySurface(EGLDisplay dpy, EGLSurface eglSurface)
         surface->wlEventQueue = NULL;
     }
 
-    for (i = 0; i < surface->ctx.numStreamImages; i++) {
-        image = surface->ctx.streamImages[i];
-        pthread_mutex_lock(&image->mutex);
+    if (surface->wlBufferEventQueue) {
+        /*
+         * If explicit sync is in use, the stream images are destroyed when
+         * destroy_surface_context() is called above.
+         */
+        WlEglStreamImage *image;
+        WlEglStreamImage *nextImage;
 
+        pthread_mutex_lock(&surface->ctx.streamImagesMutex);
         /*
          * Destroy any attached buffers to ensure no further buffer release
          * events are delivered after the buffer release queue and thread are
          * torn down.
+         *
+         * Do not destroy the WlEglStreamImages here because they may be
+         * in-flight and as soon as we unlock streamImagesMutex, the buffer
+         * release thread may access them.
          */
-        if (image->buffer) {
```

Comment:
Probably not something for this change, but I think if we made `WlEglStreamImage` refcounted, then we could simplify this teardown dance quite a lot.